src/ut_control_plane.c

```diff
@@ -262,7 +273,81 @@ static int callback_echo(struct lws *wsi, enum lws_callback_reasons reason, void
 
     return 0;
 }
+#else
+static int callback_http(struct lws *wsi, enum lws_callback_reasons reason, void *user, void *in, size_t len)
+{
+    cp_message_t msg;
+    ut_cp_instance_internal_t *pInternal = (ut_cp_instance_internal_t* )lws_context_user(lws_get_context(wsi));
+    struct per_session_data__http *pss = (struct per_session_data__http *)user;
+
+    switch (reason)
+    {
+        case LWS_CALLBACK_HTTP: {
+            UT_CONTROL_PLANE_DEBUG("LWS_CALLBACK_HTTP\n");
+            char *requested_uri = (char *)in;
+
+            if (strcmp(requested_uri, "/api/postKVP") == 0) {
+                lws_callback_on_writable(wsi);
+                return 0;
+            } else {
+                lws_serve_http_file(wsi, "index.html", "text/html", NULL, 0);
+            }
+            break;
+        }
+
+        case LWS_CALLBACK_HTTP_BODY: {
+            UT_CONTROL_PLANE_DEBUG("LWS_CALLBACK_HTTP\n");
+            if(pss != NULL)
+            {
+                UT_CONTROL_PLANE_DEBUG("LWS_CALLBACK_HTTP, pss not NULL\n");
+                if ((pss->post_data_len + len) < MAX_POST_DATA_SIZE)
+                {
+                    memcpy(pss->post_data + pss->post_data_len, in, len);
+                    pss->post_data_len += len;
+                }
+                else
+                {
+                    // POST data too large
+                    return -1;
+                }
+            }
+            break;
+        }
+
+        case LWS_CALLBACK_HTTP_BODY_COMPLETION: {
+            UT_CONTROL_PLANE_DEBUG("LWS_CALLBACK_HTTP_BODY_COMPLETION\n");
+            if(pss != NULL)
+            {
+                UT_CONTROL_PLANE_DEBUG("LWS_CALLBACK_HTTP_BODY_COMPLETION, pss not NULL\n");
+                msg.message = malloc((int)pss->post_data_len + 1);
+                assert(msg.message != NULL);
+                if (msg.message == NULL)
+                {
+                    UT_CONTROL_PLANE_ERROR("Malloc failed\n");
+                    break;
+                }
+                msg.size = (int)pss->post_data_len;
+                msg.status = DATA_RECIEVED;
+                memset(msg.message, 0, (int)len + 1);
```

Comment:
u don't need to memset it since the next thing you do is strncpy which will terminate it.